File: quant1x/data/kline_raw.go
package data

import (
	"encoding/csv"
	"fmt"
	"log"
	"os"
	"strconv"

	"gitee.com/quant1x/quant1x/quant1x/config"
)

// KLineRaw 对应 C++ 中的 data::KLineRaw，用于表示原始日线数据。
type KLineRaw struct {
	Date     string  `name:"日期" csv:"date"`        // 日期 YYYY-MM-DD
	Open     float64 `name:"开盘价" csv:"open"`       // 开盘价
	Close    float64 `name:"收盘价" csv:"close"`      // 收盘价
	High     float64 `name:"最高价" csv:"high"`       // 最高价
	Low      float64 `name:"最低价" csv:"low"`        // 最低价
	Volume   float64 `name:"成交量(股)" csv:"volume"`  // 成交量(股)
	Amount   float64 `name:"成交额(元)" csv:"amount"`  // 成交额(元)
	Up       int     `name:"上涨家数 / 外盘" csv:"up"`   // 上涨家数 / 外盘
	Down     int     `name:"下跌家数 / 内盘" csv:"down"` // 下跌家数 / 内盘
	Datetime string  `name:"时间" csv:"datetime"`    // 时间
}

//const maxKlineRawLookbackDays = 1

// SaveKlineRaw 将原始 K 线写入 CSV 文件（包含表头）。
func SaveKlineRaw(filename string, values []KLineRaw) error {
	f, err := os.Create(filename)
	if err != nil {
		return err
	}
	defer f.Close()
	w := csv.NewWriter(f)
	defer w.Flush()

	header := []string{"date", "open", "close", "high", "low", "volume", "amount", "up", "down", "datetime"}
	if err := w.Write(header); err != nil {
		return err
	}
	for _, r := range values {
		row := []string{
			r.Date,
			fmt.Sprintf("%g", r.Open),
			fmt.Sprintf("%g", r.Close),
			fmt.Sprintf("%g", r.High),
			fmt.Sprintf("%g", r.Low),
			fmt.Sprintf("%g", r.Volume),
			fmt.Sprintf("%g", r.Amount),
			fmt.Sprintf("%d", r.Up),
			fmt.Sprintf("%d", r.Down),
			r.Datetime,
		}
		if err := w.Write(row); err != nil {
			return err
		}
	}
	return nil
}

// ReadKlineRawFromCSV 从 CSV 文件中读取 KLineRaw 条目；解析错误会返回错误。
func ReadKlineRawFromCSV(filename string) ([]KLineRaw, error) {
	f, err := os.Open(filename)
	if err != nil {
		return nil, err
	}
	defer f.Close()
	r := csv.NewReader(f)
	rows, err := r.ReadAll()
	if err != nil {
		return nil, err
	}
	if len(rows) == 0 {
		return nil, nil
	}
	// 期望第一行为表头
	out := make([]KLineRaw, 0, len(rows)-1)
	for i := 1; i < len(rows); i++ {
		rec := rows[i]
		if len(rec) < 10 {
			continue
		}
		open, _ := strconv.ParseFloat(rec[1], 64)
		closev, _ := strconv.ParseFloat(rec[2], 64)
		high, _ := strconv.ParseFloat(rec[3], 64)
		low, _ := strconv.ParseFloat(rec[4], 64)
		vol, _ := strconv.ParseFloat(rec[5], 64)
		amt, _ := strconv.ParseFloat(rec[6], 64)
		up, _ := strconv.Atoi(rec[7])
		down, _ := strconv.Atoi(rec[8])
		datetime := rec[9]
		k := KLineRaw{
			Date:     rec[0],
			Open:     open,
			Close:    closev,
			High:     high,
			Low:      low,
			Volume:   vol,
			Amount:   amt,
			Up:       up,
			Down:     down,
			Datetime: datetime,
		}
		out = append(out, k)
	}
	return out, nil
}

// LoadKlineRaw 从缓存文件加载原始K线数据。
func LoadKlineRaw(code string) ([]KLineRaw, error) {
	filename := config.GetKlineFilename(code, false)
	log.Printf("[data::KLineRaw] kline file: %s", filename)
	return ReadKlineRawFromCSV(filename)
}

File: quant1x/data/kline_raw.rs
use crate::level1;
use crate::level1::protocol::Response;
use serde::{Deserialize, Serialize};
use std::thread;
use std::time::Duration;

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct KLineRaw {
    #[serde(rename = "date")]
    pub date: String,
    #[serde(rename = "open")]
    pub open: f64,
    #[serde(rename = "close")]
    pub close: f64,
    #[serde(rename = "high")]
    pub high: f64,
    #[serde(rename = "low")]
    pub low: f64,
    #[serde(rename = "volume")]
    pub volume: f64,
    #[serde(rename = "amount")]
    pub amount: f64,
    #[serde(rename = "up")]
    pub up: i32,
    #[serde(rename = "down")]
    pub down: i32,
    #[serde(rename = "datetime")]
    pub datetime: String,
}

impl KLineRaw {
    pub fn headers() -> Vec<String> {
        vec![
            "date".into(),
            "open".into(),
            "close".into(),
            "high".into(),
            "low".into(),
            "volume".into(),
            "amount".into(),
            "up".into(),
            "down".into(),
            "datetime".into(),
        ]
    }
}

// 日线最小容错回溯（偏移）天数
pub const MAX_KLINE_LOOKBACK_DAYS: usize = 1;
pub const CN_DEFAULT_TOTALFZNUM: i32 = 240; // A股默认全天交易240分钟

pub fn fetch_kline(
    code: &str,
    start: u32,
    count: u16,
    kline_type: level1::KLineType,
) -> Option<level1::SecurityBarsResponse> {
    let category = kline_type as u16;
    let start_u16 = (start.min(u16::MAX as u32)) as u16;
    let frequency = 1u16;
    let mut req =
        level1::SecurityBarsRequest::with_frequency(code, category, start_u16, count, frequency);
    let is_index = req.is_index();

    const MAX_RETRIES: usize = 3;
    const RETRY_DELAY_MS: u64 = 1000;

    for attempt in 0..=MAX_RETRIES {
        match level1::get_std_conn() {
            Ok(mut pooled) => {
                let endpoint = pooled.addr();
                let mut resp = level1::SecurityBarsResponse::new_with(is_index, category);
                match level1::protocol::process(pooled.stream(), &mut req, &mut resp)
                    .map_err(|s| std::io::Error::new(std::io::ErrorKind::Other, s))
                {
                    Ok(()) => {
                        if resp.list.is_empty() {
                            log::warn!(
                                    "[data::kline_raw] empty response from {} for {} start={} count={} cat={} zip={} unzip={} resp_count={}",
                                    endpoint,
                                    code,
                                    start,
                                    count,
                                    category,
                                    resp.header().zip_size,
                                    resp.header().unzip_size,
                                    resp.count
                                );
                        }
                        return Some(resp);
                    }
                    Err(e) => {
                        // 检查是否是连接相关错误，如果是则重试
                        let is_connection_error = e.kind() == std::io::ErrorKind::TimedOut
                            || e.kind() == std::io::ErrorKind::ConnectionRefused
                            || e.kind() == std::io::ErrorKind::ConnectionReset
                            || e.kind() == std::io::ErrorKind::ConnectionAborted
                            || e.raw_os_error() == Some(10060)  // WSAETIMEDOUT
                            || e.raw_os_error() == Some(10061)  // WSAECONNREFUSED
                            || e.raw_os_error() == Some(10054); // WSAECONNRESET

                        if is_connection_error && attempt < MAX_RETRIES {
                            log::warn!(
                                "[data::kline_raw] connection error to {} for {} start={} count={} (attempt {}/{}): {}",
                                endpoint,
                                code,
                                start,
                                count,
                                attempt + 1,
                                MAX_RETRIES + 1,
                                e
                            );
                            thread::sleep(Duration::from_millis(RETRY_DELAY_MS));
                            continue;
                        } else {
                            log::error!(
                                "[data::kline_raw] process failed for {} from {} start={} count={}: {}",
                                code,
                                endpoint,
                                start,
                                count,
                                e
                            );
                            return None;
                        }
                    }
                }
            }
            Err(e) => {
                if attempt < MAX_RETRIES {
                    log::warn!(
                        "[data::kline_raw] failed to acquire level1 client for {} (attempt {}/{}): {}",
                        code,
                        attempt + 1,
                        MAX_RETRIES + 1,
                        e
                    );
                    thread::sleep(Duration::from_millis(RETRY_DELAY_MS));
                    continue;
                } else {
                    log::error!(
                        "[data::kline_raw] failed to acquire level1 client for {}: {}",
                        code,
                        e
                    );
                    return None;
                }
            }
        }
    }

    None
}
